### ISSUE TITLE
sum_expr with a grid will not work in constraint

### ISSUE THREAD

**Issue by minmat84** [dirkschumacher/ompr#281]
Hi All

To speed up my optimization I am trying to exclude irrelevant variables x[k,j] by using a grid to define the index combinations that should enter the model. 

This works up to the point where I want to enter a constraint on the sum of the relevant x[k,jj]:

`set.seed(123)
u<- matrix(sample(c(0,0,0,0,10,12,15,55,77,95,100),9,replace=TRUE),ncol=3,nrow=3)

grid <- expand.grid(k = 1:3, j = 1:3)
grid_small <- grid[which(u>0),]  
grid_small_table <- data.table(grid_small)
grid_small_table$id <- seq.int(nrow(grid_small))

u
model<-MILPModel() %>% 
  add_variable(x[grid_small$k, grid_small$j])  %>%
  set_objective(sum_expr(u[m]*x[grid_small$k,grid_small$j], m=which(u>0)))%>%
  add_constraint(x[grid_small$k,grid_small$j] <=4)%>%
  add_constraint(x[j,k] <=3, j=1,k=3)%>%
  add_constraint(sum_expr(x[grid_small$k,grid_small$j])<=9)%>%
  solve_model(with_ROI("glpk",verbose=TRUE))
  get_solution(model, x[k,j])
    
model$objective_value  `

But this will give me an error

`
Error in FUN(X[[i]], ...) : not implemented`

 I also tried to implement it in the same way as the objective function, by doing 
this: 

`one <- rep(1,3*3)
add_constraint(sum_expr(one[m]*x[grid_small$k,grid_small$j],m=which(u>0))<=9)`

But then
`<SOLVER MSG>  ----
GLPK Simplex Optimizer, v4.47
15 rows, 7 columns, 15 non-zeros
*     0: obj =  0.000000000e+000  infeas = 0.000e+000 (0)
*     7: obj =  1.604000000e+003  infeas = 0.000e+000 (0)
OPTIMAL SOLUTION FOUND
<!SOLVER MSG> ----
>   get_solution(model, x[k,j])
  variable k j value
1        x 2 1     4
2        x 3 1     4
3        x 1 2     4
4        x 2 2     4
5        x 1 3     3
6        x 2 3     4
7        x 3 3     4
>   
>   
> model$objective_value
[1] 1604`

The constraint that the sum of the x should be lower or equal than 9 will be ignored. Any ideas?

**Comment by hugolarzabal** [comment]
Hello,

There were several problems in your example:

- You need to specify a lower bound for your variable or it may take negative values.
- You need to specify that the sense of the objectiv eis to maximize (if not all x = 0 is the best solution)
- I assume you also want x to be an integer with type = "integer"

Then about your problem came from the `sum_expr` function.
The sum_expr function is a bit unintuitive and doesn't work the same as a normal constraint function. This is because sum_expr uses expand.grid on every vector passed to it (unless it is colwise).

Luckily, sum_expr also ignore every combination of indices which do not exist (and send a warning about it)

So in sum_expr, you need to:

- pass every values of k and j
- use colwise() on the vector containing the values of your parameter u
`set_objective(sum_expr(colwise(grid_small$u) * x[k,j], k = 1:3, j=1:3), sense = "max")`

The entire code is the following:
```
set.seed(123)
u<- matrix(sample(c(0,0,0,0,10,12,15,55,77,95,100),9,replace=TRUE),ncol=3,nrow=3)

grid <- expand.grid(k = 1:3, j = 1:3)
grid_u <- grid %>% mutate(u = test(k,j))
grid_small <- grid[which(u>0),]
grid_small$u <- u[which(u>0)]

model<- MILPModel() %>%
  add_variable(x[grid_small$k, grid_small$j], type = "integer", lb = 0) %>%
  set_objective(sum_expr(colwise(grid_small$u) * x[k,j], k = 1:3, j=1:3), sense = "max") %>%
  add_constraint(x[grid_small$k,grid_small$j] <=4)%>%
  add_constraint(x[1,3] <=3) %>%
  add_constraint(sum_expr(x[k, j], k = 1:3, j=1:3)<=9)
  
sol <- model %>% solve_model(with_ROI("glpk",verbose=TRUE))

get_solution(sol, x[k,j])
```

**Comment by minmat84** [comment]
Hello @hugolarzabal 

thank you very much for looking into my problem.

Still, it won't accept me to feed in the x[k,j] with non-existing indices and sends an error, no warning: 

` model<- MILPModel() %>%
   add_variable(x[grid_small$k, grid_small$j], type = "integer", lb = 0) %>%
   set_objective(sum_expr(colwise(grid_small$u) * x[k,j], k = 1:3, j=1:3), sense = "max") %>%
  add_constraint(x[grid_small$k,grid_small$j] <=4) %>%
   add_constraint(x[1,3] <=3) %>%
   add_constraint(sum_expr(x[k, j], k = 1:3, j=1:3)<=9)`
`Error: You used the variable 'x' with at least one index that does not exists.`

I am really having a hard time to debug my code as I cannot explicitely evaluate sum_expr.

**Comment by minmat84** [comment]
So is there a way to directly prevent index combinations to enter the model explicitely ? 

If not this would be a great functionality to be implemented some day...

**Comment by hugolarzabal** [comment]
Hello,

When I execute your code, I do not get an error, only a warning:

> Warning messages:
> 1: You used the variable 'x' with 9 indexes but only 7 indexes will be used 
> 2: You used the variable 'x' with 9 indexes but only 7 indexes will be used 

What version of ompr are you using? The difference may come from there.

**Comment by minmat84** [comment]
Helloy @hugolarzabal 

I am using ompr 0.8:

```
R version 3.4.1 (2017-06-30)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=German_Switzerland.1252  LC_CTYPE=German_Switzerland.1252    LC_MONETARY=German_Switzerland.1252
[4] LC_NUMERIC=C                        LC_TIME=German_Switzerland.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bindrcpp_0.2.2        ompr.roi_0.8.0        ROI.plugin.glpk_0.3-0 Rglpk_0.6-4          
[5] slam_0.1-44           ROI_0.3-2             ompr_0.8.0            dplyr_0.7.8          

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0                ROI.plugin.symphony_0.2-5 lpSolveAPI_5.5.2.0-17     bindr_0.1.1              
 [5] magrittr_1.5              tidyselect_0.2.5          lattice_0.20-35           R6_2.2.2                 
 [9] rlang_0.3.1               ROI.plugin.clp_0.4        tools_3.4.1               grid_3.4.1               
[13] data.table_1.12.0         clpAPI_1.2.9              registry_0.5              lazyeval_0.2.1           
[17] assertthat_0.2.0          tibble_2.0.1              crayon_1.3.4              Matrix_1.2-10            
[21] purrr_0.2.5               ROI.plugin.lpsolve_0.3-2  glue_1.3.0                Rsymphony_0.1-28         
[25] compiler_3.4.1            pillar_1.3.1              pkgconfig_2.0.2 
```

**Comment by hugolarzabal** [comment]
Hello,
You should download the last version of ompr from github.
The sum_expr warning instead of error has been implemented about one year ago but version 0.8.0 is older than that.

**Comment by dirkschumacher** [comment]
You might want to take a look at the most recent `master` version and try the `MIPModel`. I close the issue for now, but feel free to open a new one.
